storage/innobase/log/log0online.cc

```diff
@@ -1335,24 +1518,29 @@ bool log_online_follow_redo_log_one_pass(void) {
 
   ut_ad(!log_bmp_sys->out.file.is_closed());
 
-  /* Grab the LSN of the last checkpoint, we will parse up to it */
-  /* Parse up to the LSN of the last checkpoint */
-  log_bmp_sys->follow_up_to(log_get_checkpoint_lsn(*log_sys));
+  // Grab the LSN of the last checkpoint. It is possible that checkpoint LSN
+  // points to the middle of a log record. We'd need to find next record group
+  // beginning (which is possibly located in one of the following log blocks)
+  // and parse data up to that point.
+  log_bmp_sys->set_checkpoint_lsn(log_get_checkpoint_lsn(*log_sys));
+
+  ut_ad(log_bmp_sys->parse_buf.get_parse_start_lsn() == 0);
 
-  if (log_bmp_sys->end_lsn == log_bmp_sys->start_lsn) {
+  if (log_bmp_sys->checkpoint_lsn <= log_bmp_sys->has_parse_data_to()) {
```

Comment:
what is the difference  bmp_sys->start_lsn and has_parse_data_to()?

Reply:
The has_parse_data_to() checks last known LSN in parse buffer at first place. It may return bmp_sys->start_lsn as a fallback in case there were not any data parsed yet.